### ISSUE TITLE
[bug] Generic provenance generator requires contents: write even if upload-assets: false

### ISSUE THREAD

**Issue by btj** [slsa-framework/slsa-github-generator#2044]
The generic provenance generator fails if it does not get `contents: write` permission, even if `upload-assets: false` is specified. I am not comfortable giving you guys write access to my repo.

**Comment by ianlewis** [comment]
Thanks for the feedback. Unfortunately we can't dynamically request token permissions from GitHub so we're currently limited by GitHub Actions to request all permissions that the workflow might need.

Current options include:
1. Create separate workflows: one that supports uploading provenance to releases, and one that doesn't. Each workflow could require separate permissions.
2. Remove the feature for uploading to releases.

**Comment by ianlewis** [comment]
Related #1257

**Comment by laurentsimon** [comment]
@btj we've been talking about creating multiple workflows due to the limitations of GitHub reusable workflows. Maybe it's about time we follow through. I was hoping GitHub would update their re-usable workflow to dynamically resolved permissions.

/cc @josepalafox, @feelepxyz @bdehamer @kommendorkapten

**Comment by ianlewis** [comment]
We have received some feedback wrt the Node.js builder that folks want to separate the "build" and "publish" steps rather than having the reusable workflow take care of everything. We may want to reconsider whether we need to actually do the "publish" at all. I think right now it's just a UX thing that makes it easier for some users but isn't a necessary step.

**Comment by chrispat** [comment]
> @btj we've been talking about creating multiple workflows due to the limitations of GitHub reusable workflows. Maybe it's about time we follow through. I was hoping GitHub would update their re-usable workflow to dynamically resolve permissions. Their current implementation does _not_ follow the least privilege principles; as it forces callers to provide more permissions than actually needed.

The current implementation requires the caller to provide the maximum permissions they want to allow the called workflow to have.  Within that workflow you can downgrade the permissions for any individual job but not upgrade.

**Comment by laurentsimon** [comment]
Yes we know that. But that still relies on trusting the re-usable workflow fully. The threat model here is that users want to minimize the number of permissions they pass to a re-usable workflow, based on their use cases. By design, re-usable workflows are used to to pack a lot of different features, which means they accumulate more and more permissions over time. Users should still be able to decide which of these features they want to give the re-usable workflow access to. I think this is important to reduce supply-chain risks.

This is similar to smartphone permissions. I'm fine giving WhatsApp access to my images, but maybe not my locations - even though the WhatsApp request all of them.
